doc/dev/cephadm: document the cephadm exporter design (#38329)
ceph/ceph: doc/dev/cephadm/cephadm-exporter.rst

Additional ceph commands
------------------------
::

# ceph cephadm generate-exporter-config

This command will create generate a default configuration consisting of; a self signed certificate, a randomly generated
32 character token and the default port of 9443 for the REST API.
::

# ceph cephadm set-exporter-config -i <config.json>

Use a JSON file to define the crt, key, token and port for the REST API. The crt, key and token are validated by
the mgr/cephadm module prior storing the values in the KV store. Invalid or missing entries should be reported to the 
user.
::

# ceph cephadm clear-exporter-config

Clear the current configuration (removes the associated keys from the KV store)
::

# ceph cephadm get-exporter-config

Show the current exporter configuration, in JSON format

.. note:: If the service is already deployed any attempt to change or clear the configuration will
    be denied. In order to change settings you must remove the service, apply the required configuration
    and re-apply (``ceph orch apply cephadm-exporter``)

New Ceph Configuration Keys
===========================

> The exporter configuration is persisted to the monitor's KV store, as the following settings;

Reviewer: ```suggestion
The exporter configuration is persisted to the monitor's KV store, as the following settings:
```

    
    print("Reading config.json")
    try:
        with open('./config.json', 'r') as f:
            raw=f.read()
    except FileNotFoundError as e:
        print("You must first create a config.json file using the cephadm get-exporter-config command")
        sys.exit(1)

    cfg = json.loads(raw)
    with tempfile.NamedTemporaryFile(buffering=0) as t:
        print("creating a temporary local crt file from the json")
        t.write(cfg['crt'].encode('utf-8'))

        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.load_verify_locations(t.name)
        hdrs={"Authorization":f"Bearer {cfg['token']}"}
        print("Issuing call to gather metadata")
        req=Request(f"https://{hostname}:9443/v1/metadata",headers=hdrs)
        s_time = time.time()
        r = urlopen(req,context=ctx)
        print(r.status)
        print("call complete")
        # assert r.status == 200
        if r.status in [200, 206]:

            raw=r.read()  # bytes string
            js=json.loads(raw.decode())
            print(json.dumps(js, indent=2))
        elapsed = time.time() - s_time
        print(f"Elapsed secs : {elapsed}")

> .. note:: the above example is using python3, and assumes that you've extracted the config using the get-exporter-config command.

Reviewer: ```suggestion
.. note:: the above example uses python3 and assumes that you've extracted the config using the ``get-exporter-config`` command.
```

Author: ack

Implementation Specific Details
===============================

> Like a typical container based deployment, the exporter is deployed to a directory under ``/var/lib/ceph/<fsid>``. The 

Reviewer: ```suggestion
The exporter is deployed to a directory under ``/var/lib/ceph/<fsid>``, as in a typical container-based deployment. The 
```

Author: re-phrased. thanks

place over a simple GET interface. Although only the GET method is supported, the API provides multiple URLs to
provide different views on the metadata that has been gathered.

.. csv-table:: Supported URL endpoints
    :header: "URL", "Purpose"

    "/v1/metadata", "show all metadata including health of all threads"
    "/v1/metadata/health", "only report on the health of the data gathering threads"
    "/v1/metadata/disks", "show the disk output (ceph-volume inventory data)"
    "/v1/metadata/host", "show host related metadata from the gather-facts command"
    "/v1/metatdata/daemons", "show the status of all ceph cluster related daemons on the host"

Return Codes
------------
The following HTTP return codes are generated by the API

.. csv-table:: Supported HTTP Responses
    :header: "Status Code", "Meaning"

    "200", "OK"
    "204", "the thread associated with this request is no longer active, no data is returned"
    "206", "some threads have stopped, so some content is missing"
    "401", "request is not authorised - check your token is correct"
    "404", "URL is malformed, not found"

>     "500", "all threads have stopped - unable to provide any metadata for the host"

Reviewer 2: Isn't this already documented by saying we're using a REST API? Sounds duplicated to list standard HTTP codes here.

Author: yeah - the codes are standard, as you'd expect...but what I'm documenting here is the meaning wrt how the exporter is working or perhaps not working :)
